op-batcher/batcher/espresso/streamer.go

```diff
@@ -0,0 +1,225 @@
+	"C"
+	"cmp"
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"slices"
+
+	espressoClient "github.com/EspressoSystems/espresso-network-go/client"
+	espressoLightClient "github.com/EspressoSystems/espresso-network-go/light-client"
+	espressoTypes "github.com/EspressoSystems/espresso-network-go/types"
+	espressoVerification "github.com/EspressoSystems/espresso-network-go/verification"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// espresso-network-go's HeaderInterface currently lacks a function to get this info,
+// although it is present in all header versions
+func getFinalizedL1(header *espressoTypes.HeaderImpl) *espressoTypes.L1BlockInfo {
+	v0_1, ok := header.Header.(*espressoTypes.Header0_1)
+	if ok {
+		return v0_1.L1Finalized
+	}
+	v0_2, ok := header.Header.(*espressoTypes.Header0_2)
+	if ok {
+		return v0_2.L1Finalized
+	}
+	v0_3, ok := header.Header.(*espressoTypes.Header0_3)
+	if ok {
+		return v0_3.L1Finalized
+	}
+	return nil
+}
+
+type L1Client interface {
+	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
+}
+
+type EspressoStreamer struct {
+	// Namespace of the rollup we're interested in
+	Namespace uint64
+	// Address of the batcher, we expect transactions to
+	// be signed by the corresponding private key
+	BatcherAddress common.Address
+
+	L1Client            L1Client
+	EspressoClient      *espressoClient.Client
+	EspressoLightClient *espressoLightClient.LightClientReader
+	Log                 log.Logger
+
+	// Batch number we're to give out next
+	BatchPos uint64
+	// Position of ???
+	confirmedBatchPos uint64
+	// HotShot block we're to fetch next
+	hotShotPos uint64
+
+	// Maintained in sorted order, but may be missing batches if we receive
+	// any out of order.
+	batchBuffer []EspressoBatch
+}
+
+// Handle both L1 reorgs and batcher restarts by updating our state in case it is
+// not consistent with what's on the L1. Returns true if the state was updated.
+func (s *EspressoStreamer) Refresh(ctx context.Context, syncStatus *eth.SyncStatus) (bool, error) {
+	if s.confirmedBatchPos == syncStatus.SafeL2.Number {
+		return false, nil
+	}
+
+	hotshotState, err := s.EspressoLightClient.LightClient.
+		FinalizedState(&bind.CallOpts{BlockNumber: new(big.Int).SetUint64(syncStatus.SafeL2.L1Origin.Number)})
+	if err != nil {
+		return false, err
+	}
+
+	s.confirmedBatchPos = syncStatus.SafeL2.Number
+	s.BatchPos = s.confirmedBatchPos + 1
+	s.hotShotPos = hotshotState.BlockHeight
+	s.batchBuffer = nil
+	return true, nil
+}
+
+func (s *EspressoStreamer) Update(ctx context.Context) error {
+	// Fetch more batches from HotShot if available.
+	hotshotState, err := s.EspressoLightClient.LightClient.FinalizedState(&bind.CallOpts{})
```

Comment:
Flagging that we shouldn't be reading from the light client on the critical path of fast confirmations, as it might only be updated every 10 minutes or less. But I think there is already WIP to switch this to querying the Espresso nodes directly, if so ignore this comment